symbolic-base/src/ZkFold/Base/Data/Conditional.hs

```diff
@@ -0,0 +1,100 @@
+{-# LANGUAGE DerivingVia          #-}
+{-# LANGUAGE TypeOperators        #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module ZkFold.Base.Data.Conditional
+  ( Conditional (..)
+  , GConditional (..)
+  ) where
+
+import qualified Data.Bool                        as H
+import           Data.Functor.Identity
+import qualified Data.Functor.Rep                 as R
+import           Data.Kind
+import qualified GHC.Generics                     as G
+import           Prelude                          (type (~))
+
+import           ZkFold.Base.Algebra.Basic.Number
+import           ZkFold.Base.Data.Boolean
+import           ZkFold.Base.Data.Vector
+
+class Boolean (BooleanOf a) => Conditional a where
+
+  type BooleanOf a :: Type
+  type BooleanOf a = GBooleanOf (G.Rep a)
+
+  ifThenElse :: BooleanOf a -> a -> a -> a
+  default ifThenElse ::
+    ( G.Generic a
+    , GConditional (G.Rep a)
+    , BooleanOf a ~ GBooleanOf (G.Rep a)
+    ) => BooleanOf a -> a -> a -> a
+  ifThenElse b t f = G.to (gifThenElse b (G.from t) (G.from f))
+
+instance
+  ( Conditional x0
+  , Conditional x1
+  , BooleanOf x0 ~ BooleanOf x1
+  ) => Conditional (x0,x1)
+
+instance
+  ( Conditional x0
+  , Conditional x1
+  , Conditional x2
+  , BooleanOf x0 ~ BooleanOf x1
+  , BooleanOf x1 ~ BooleanOf x2
+  ) => Conditional (x0,x1,x2)
+
+instance
+  ( Conditional x0
+  , Conditional x1
+  , Conditional x2
+  , Conditional x3
+  , BooleanOf x0 ~ BooleanOf x1
+  , BooleanOf x1 ~ BooleanOf x2
+  , BooleanOf x2 ~ BooleanOf x3
+  ) => Conditional (x0,x1,x2,x3)
+
+instance (R.Representable v, Conditional x)
+  => Conditional (G.Generically1 v x) where
+    type BooleanOf (G.Generically1 v x) = BooleanOf x
+    ifThenElse b (G.Generically1 vt) (G.Generically1 vf) =
+      G.Generically1 (R.mzipWithRep (ifThenElse b) vt vf)
+
+deriving via G.Generically1 (Vector n) x
+  instance (KnownNat n, Conditional x) => Conditional (Vector n x)
```

Comment:
If we intend to use this instance in a generic context, then it is not as efficient as can be in a usual Haskell context as this would require $O(n)$ instead of the standard $O(1)$

Reply:
Don't know what to do here right now, I have the following idea but I don't know if this looks clean to you
```haskell
class Boolean b where
  ...
  choice :: (a -> a -> b -> a) -> (e -> a) -> (e -> a) -> b -> (e -> a)
  -- chooses a closure for @b ~ base.Data.Bool@, creates new closure for @b ~ ZkFold.Symbolic.Data.Bool@
```

Reply:
(Terminology borrowed from [here](https://github.com/JetBrains/arend-lib/blob/master/src/Logic/Classical.ard#L12) as this looks familiar)

Reply:
OR we could define two overlapping instances for `Vector`, but this would require the user to propagate the `Conditional (Vector n x)` constraint until `BooleanOf x` is known to be either `base.Data.Bool` or not

Reply:
OR we could define this `choice` to have a type `(Conditional a, BooleanOf a ~ b) => Choose (e -> a) b`, but this is not really better looking and makes `Boolean` and `Conditional` depend on each other unless we drop the dependency on `Boolean` in `Conditional`, but I don't like this approach

Reply:
OR we can ultimately decide that the sort of types which are `Conditional` on a certain boolean type `b` have to be decided by `b` itself, not by the types in sort, like
```haskell
class Boolean b => Conditional b where
  type Universe b :: Type -> Constraint
  ifThenElse :: Universe b a => b -> a -> a -> a
```
(name `Universe` is up to debate, of course, as this name is too generic)
Then we could have:
* `Universe Prelude.Bool` to be some universally satisfied constraint (already appeared in my [collection of thoughts on numerical hierarchy](https://gist.github.com/TurtlePU/55e0f3a8ff0929be84199bcf643de58e#file-poly-hs-L155), btw)
* `Universe (Bool ctx) a` to be `(SymbolicData a, Context a ~ ctx)` (this would require a trick with a phantom class too, though)

Reply:
Sorry that I didn't notice this earlier, only after witnessing this `Vector` instance x)

Reply:
This same objection holds for `Conditional` instances derived for generic Haskell products, which will be less efficient than Haskell's `bool` or standard `if then else`, which is of course _still available_ for all types, Haskell or Symbolic. The change to `Conditional` you suggest is a major one.

Reply:
Oh, indeed...

Reply:
Here is my reasoning. For `x` such that `SymbolicData x`, there is only one choice of boolean: `Bool (Context x)`. So, ideally, there should be only one choice of `Eq` and `Conditional` for `x`.

Maybe we could override the generic implementation in the case of `Interpreter`? Would this solve the efficiency issue?

Regarding the elliptic curve hierarchy... The situation is different. As I noted, we have several options how the `BaseField` of the context can be related to the elliptic curve point type. Depending on the field choice, the implementation of the EC operations will be different. So, a multi-parametric approach is justified (and it is very much needed in the IVC, where we want to use all those options).

Reply:
> So, ideally, there should be only one choice of `Eq` and `Conditional` for `x`.

This would be of course ideal, but, for `Conditional`, this is not really optimal -- for `instance Conditional (Vector a)` we have the following dilemma:

* if we keep only one instance, then this should be an instance which builds new vector componentwise, as done in the PR. This is suboptimal in pure Haskell context; and we would still need instance for pure Haskell if we want to write down language-agnostic generalized algorithms, which is the point of @echatav's refactoring.

* if we need more instances, then we need a multiparam class or something of the sort.

Actually, I think the current multiparam class definition is fine, it's the instances that have to be changed to:

```haskell
class BoolType b => Conditional b a where
  ifThenElse :: b -> a -> a -> a

instance Conditional Haskell.Bool a where
  ifThenElse condition thenBranch elseBranch =
    if condition then thenBranch else elseBranch

instance SymbolicData a => Conditional (Bool (Context a)) a where
  ifThenElse (Bool c) t e = unPar1 $ interpolate [(zero, e), (one, t)] c
```

Where `interpolate` is a member of `SymbolicData` with the following type:
```haskell
class SymbolicData a where
  ...
  interpolate :: (Context a ~ c, Representable f) => [(BaseField c, a)] -> c f -> f a
```
(For now, `interpolate` can be defined only on lists of size 2 for simplicity of implementation)

The reasoning behind `interpolate` is such that once we have sum types in Symbolic, we will need to define an eliminator for them. For `Bool`, this is `Conditional`; to avoid defining a separate class for each sum type, I think it's better to include the common functionality in the base `SymbolicData` class and call into it.

Reply:
> return to the design of having `Conditional` and `Eq` being multiparameter typeclasses

Agreed about `Conditional` being multiparam, but I think `Eq` should still have an associated type, like this:
```
class Conditional (BooleanOf a) a => Eq a where
  type BooleanOf a :: Type
  (==) :: a -> a -> BooleanOf a
  (/=) :: a -> a -> BooleanOf a
```
My arguments are:
* Runtime performance doesn't differ much between `Haskell.Eq` and this `Eq`, even in pure context
* This makes _sense_ because while `Conditional` is like an eliminator for a boolean datatype, `Eq a` is a property of `a` that its equality relation is decidable (well, what "decidable" means in Symbolic is an interesting question, but still).
* On the topic of type inference. In the case of `Conditional`, having it multiparam is fine because all the types can be inferred from the inputs. Output of `Eq`, on the other hand, cannot be inferred from the inputs directly so we have to provide some help to inference algorithm in form of either fundeps or associated types.

Reply:
^ Yes, I agree that having its boolean be inferable from a type makes more sense for `Eq` than `Conditional`.